### PR TITLE
checkConn for badConn and alwaysNewConn(#2053)

### PR DESCRIPTION
**Pull request by naiqianz**
#2053
check bad conn when getConn from pool and add maxBadConnRetries

### CODE REVIEW

**Comment by monkey92t** [comment]
It seems that the issue of #1849 was not resolved.

**Comment by monkey92t** [comment]
It seems that we should consider the database/sql solution(https://github.com/go-redis/redis/issues/1737#issuecomment-832434567):

```
func (db *DB) ExecContext(ctx context.Context, query string, args ...interface{}) (Result, error) {
	var res Result
	var err error
	for i := 0; i < maxBadConnRetries; i++ {
		res, err = db.exec(ctx, query, args, cachedOrNewConn)
		if err != driver.ErrBadConn {
			break
		}
	}
	if err == driver.ErrBadConn {
		return db.exec(ctx, query, args, alwaysNewConn)
	}
	return res, err
}
```

**Comment by monkey92t** [comment]
@vmihailenco  Do you have any comments?

**Comment by monkey92t** [comment]
WechatIMG97

File: internal/pool/conn_check.go [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+	"net"
+	"syscall"
+	"time"
+)
+
+var errUnexpectedRead = errors.New("unexpected read from socket")
+
+func connCheck(conn net.Conn) error {
+	//reset deadline
+	_ = conn.SetDeadline(time.Time{})
```
Comment (vmihailenco) [inline internal/pool/conn_check.go:17]: @monkey92t I believe this line resolves https://github.com/go-redis/redis/pull/1849/files. I've been using similar code with go-clickhouse and everything seems to be working fine.

Comment (vmihailenco) [inline internal/pool/conn_check.go:17]: @naiqianz could you fix the comment formatting?

```go
// Reset previous timeout.
```

File: internal/pool/pool.go [hunk 2]
```diff
@@ -250,7 +255,7 @@ func (p *ConnPool) Get(ctx context.Context) (*Conn, error) {
 		return nil, err
 	}
 
-	for {
+	for i := 0; i < maxBadConnRetries; i++ {
```
Comment (vmihailenco) [inline internal/pool/pool.go:258]: Let's split this change in a separate PR and leave only checkConn like it was done in https://github.com/go-redis/redis/issues/1821

**Comment by vmihailenco** [comment]
>It seems that we should consider the database/sql solution

We probably need both this change and database/sql-like retry approach. `checkConn` is about early detection of bad connections and we need it anyway.

File: internal/pool/conn_check.go (hunk 1 above)
Comment (naiqianz) [reply internal/pool/conn_check.go:17]: sure, tks for ur suggestion

**Comment by monkey92t** [comment]
We once made a mistake in `SetDeadline` and should add a check for it.

**Comment by monkey92t** [comment]
It looks good, we will add `maxBadConnRetries`, `cachedOrNewConn `, `alwaysNewConn ` later.

**Review by monkey92t — APPROVED** [review]
